test/local/routes/signin-utils.js

```diff
@@ -0,0 +1,128 @@
+        .then(matches => {
+          assert.ok(matches, 'password matches, checkPassword returns true')
+          assert.equal(MockCustoms.flag.callCount, 0, 'customs.flag was not called')
+        })
+    }
+  )
+
+  it(
+    'should return false when check with incorrect password',
+    () => {
+      const authPW = Buffer.from('aaaaaaaaaaaaaaaa')
+      const emailRecord = {
+        uid: 'uid',
+        email: 'test@example.com',
+        verifyHash: null,
+        verifierVersion: 0,
+        authSalt: Buffer.from('bbbbbbbbbbbbbbbb')
+      }
+      MockCustoms.flag.reset()
+
+      const password = new Password(
+              authPW, emailRecord.authSalt, emailRecord.verifierVersion)
+
+      return password.verifyHash()
+        .then(hash => {
+          emailRecord.verifyHash = hash
+          const incorrectAuthPW = Buffer.from('cccccccccccccccc')
+          const incorrectPassword = new Password(
+              incorrectAuthPW, emailRecord.authSalt, emailRecord.verifierVersion)
+          return signinUtils.checkPassword(emailRecord, incorrectPassword, CLIENT_ADDRESS)
+        })
+        .then(match => {
+          assert.equal(!! match, false, 'password does not match, checkPassword returns false')
+          assert.equal(MockCustoms.flag.callCount, 1, 'customs.flag was called')
+          assert.equal(MockCustoms.flag.getCall(0).args[0], CLIENT_ADDRESS, 'customs.flag was called with client ip')
+          assert.deepEqual(MockCustoms.flag.getCall(0).args[1], {
+            email: emailRecord.email,
+            errno: error.ERRNO.INCORRECT_PASSWORD
+          }, 'customs.flag was called with correct event details')
+        })
+    }
+  )
+
+  it(
+    'should error when check with account whose password must be reset',
+    () => {
+      const emailRecord = {
+        uid: 'must_reset',
+        email: 'test@example.com',
+        verifyHash: null,
+        verifierVersion: 0,
+        authSalt: butil.ONES
+      }
+      MockCustoms.flag.reset()
+
+      const incorrectAuthPW = Buffer.from('cccccccccccccccccccccccccccccccc')
+      const incorrectPassword = new Password(
+          incorrectAuthPW, emailRecord.authSalt, emailRecord.verifierVersion)
+
+      return signinUtils.checkPassword(emailRecord, incorrectPassword, CLIENT_ADDRESS)
+        .then(
+          (match) => { assert(false, 'password check should not have succeeded') },
+          (err) => {
+            assert.equal(err.errno, error.ERRNO.ACCOUNT_RESET, 'an ACCOUNT_RESET error was thrown')
+            assert.equal(MockCustoms.flag.callCount, 1, 'customs.flag was called')
+            assert.equal(MockCustoms.flag.getCall(0).args[0], CLIENT_ADDRESS, 'customs.flag was called with client ip')
+            assert.deepEqual(MockCustoms.flag.getCall(0).args[1], {
+              email: emailRecord.email,
+              errno: error.ERRNO.ACCOUNT_RESET
+            }, 'customs.flag was called with correct event details')
+          }
+        )
+    }
+  )
+})
```

Comment:
It would probably be a good idea for this file to acquire tests for each of the new helper methods in the signin-utils module.  Currently those helpers are exercised by the existing login-related tests in `./test/local/routes/account.js`, and they should be refactored to live in here instead of in there.

I'm happy to take that on if you think it's worthwhile as part of this PR, but I didn't want to complicate what's already a bit diff...

Reply:
I'm not going to insist on anything, but if it was me I think I'd refactor the tests now as part of this PR. Those refactorings tend not to end up happening if left for later, in my experience. (but maybe that's just me)

And there's definitely value in moving tests down to the appropriate level imho, it's much easier to test individual conditions (and thus uncover any small bugs) when you're not going via an intermediary.

And although leaving it until later makes for a smaller review, it also makes for a disjointed one, where the reviewer doesn't have all of the context paged into their working set.

But those are just opinions, so 100% your call.

Reply:
(and also they're uninformed opinions, stated without any understanding of how complicated said refactoring is going to be in practice)

Reply:
> Those refactorings tend not to end up happening if left for later, in my experience

You're right, that's broadly my experience as well.  It's probably worth taking the time to do it while the context is all still fresh, and there's probably a good chance it'll uncover at least one bug :-)